### PR TITLE
Remove tmpnam from tests

### LINKED ISSUE #1493: the use of `tmpnam' is dangerous, better use `mkstemp'
Fix this annoying warning for good ?

### PR DESCRIPTION
**Pull request by payetvin**

### CODE REVIEW

**Review by flomnes — COMMENTED** [review]
The behavior of `std::filesystem::temp_directory_path` is not strictly equivalent.

You just removed a call to `tmpnam`, which is guaranteed to create a **unique** directory, avoiding possible side-effects (a file that wasn't removed by the previous test makes this one succeed/fail).

**Review by flomnes — CHANGES_REQUESTED** [review]
Also, /tmp already exists, no need to create it

File: src/tests/src/libs/antares/study/constraint/test_constraint.cpp [hunk 1]
```diff
@@ -15,6 +15,15 @@
...
+fs::path generateAndCreateDirName()
+{
+    std::string dir = boost::unit_test::framework::current_test_case().p_name;
+    fs::path working_dir = fs::temp_directory_path() / dir;
+    fs::remove(working_dir);
+    fs::create_directories(working_dir);
+    return working_dir;
+}
+
 BOOST_AUTO_TEST_SUITE(BindingConstraintTests)
```
Comment (flomnes) [inline src/tests/src/libs/antares/study/constraint/test_constraint.cpp]: You may need to remove unwanted characters, not all characters are valid in a path

Also, please use a single declaration/definition

Comment (payetvin) [reply src/tests/src/libs/antares/study/constraint/test_constraint.cpp]: I don't see which unwanted characters could be found, we're only using alpha numeric chars and underscores in test names.

This function is used in different places, should it be moved in a header at the root of src/tests ?

Comment (flomnes) [reply src/tests/src/libs/antares/study/constraint/test_constraint.cpp]: src/tests/end-to-end/utils/utils.h

File: src/tests/src/solver/simulation/test-time_series.cpp [hunk 2]
```diff
@@ -43,9 +44,8 @@ struct Fixture {
     Fixture() {
         study = std::make_shared<Study>();
         study->header.version = version870;
-        auto tmp_dir = fs::temp_directory_path();
-        working_tmp_dir = tmp_dir / std::tmpnam(nullptr);
-        fs::create_directories(working_tmp_dir);
+        std::string testName = boost::unit_test::framework::current_test_case().p_name;
+        working_tmp_dir = generateAndCreateDirName(testName);
```
Comment (flomnes) [inline src/tests/src/solver/simulation/test-time_series.cpp]: You could create this macro in utils.h / utils.cpp

File: src/tests/src/libs/antares/study/constraint/test_constraint.cpp [hunk 3]
```diff
@@ -5,12 +5,15 @@
...
 #define BOOST_TEST_MODULE binding_constraints
 
-#include <boost/test/included/unit_test.hpp>
+#include <boost/test/unit_test.hpp>
 #include <fstream>
 #include "antares/study/constraint.h"
 #include "antares/study/area/area.h"
 #include "antares/study.h"
 #include <filesystem>
+#include "../../../../utils/utils.h"
```
Comment (flomnes) [inline src/tests/src/libs/antares/study/constraint/test_constraint.cpp]: You should not need to use "../../../"

File: src/tests/src/libs/antares/study/constraint/test_constraint.cpp [hunk 4]
```diff
@@ -5,12 +5,15 @@
...
+
+#define testName boost::unit_test::framework::current_test_case().p_name
```
Comment (flomnes) [inline src/tests/src/libs/antares/study/constraint/test_constraint.cpp]: Define in utils.h ? Or even better
```cpp
CREATE_TMP_DIR_BASED_ON_BOOST_TEST_NAME();
```

**Review by flomnes — CHANGES_REQUESTED** [review]
I believe that utils.h/.cpp already exist in `develop`, please take a look

File: src/tests/src/libs/antares/study/constraint/test_constraint.cpp (hunk 3 above)
Comment (payetvin) [reply src/tests/src/libs/antares/study/constraint/test_constraint.cpp]: Fixed

File: src/tests/src/libs/antares/study/constraint/test_constraint.cpp [hunk 5]
```diff
@@ -5,12 +5,13 @@
 #define WIN32_LEAN_AND_MEAN
 #define BOOST_TEST_MODULE binding_constraints
 
-#include <boost/test/included/unit_test.hpp>
+#include <boost/test/unit_test.hpp>
```
Comment (flomnes) [inline src/tests/src/libs/antares/study/constraint/test_constraint.cpp:10]: Please add comment here 
// Why this header ? See here https://www.boost.org/doc/libs/1_82_0/libs/test/doc/html/boost_test/adv_scenarios/single_header_customizations/multiple_translation_units.html

File: src/tests/src/utils/utils.h [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <filesystem>
+
+#define CREATE_TMP_DIR_BASED_ON_TEST_NAME() generateAndCreateDirName(boost::unit_test::framework::current_test_case().p_name);
+
+std::filesystem::path generateAndCreateDirName(std::string);
```
Comment (flomnes) [inline src/tests/src/utils/utils.h]:
```suggestion
std::filesystem::path generateAndCreateDirName(const std::string&);
```

File: src/tests/src/utils/utils.cpp [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#include "utils.h"
+
+namespace fs = std::filesystem;
+
+fs::path generateAndCreateDirName(std::string dirName)
```
Comment (flomnes) [inline src/tests/src/utils/utils.cpp]:
```suggestion
fs::path generateAndCreateDirName(const std::string& dirName)
```

File: src/tests/src/utils/utils.h [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <filesystem>
+
+#define CREATE_TMP_DIR_BASED_ON_TEST_NAME() generateAndCreateDirName(boost::unit_test::framework::current_test_case().p_name);
```
Comment (flomnes) [inline src/tests/src/utils/utils.h:6]: Add a comment here to explain why we need a macro (linking issues with boost)

**Review by flomnes — CHANGES_REQUESTED** [review]

**Review by flomnes — CHANGES_REQUESTED** [review]
Merge **src/tests/src/utils** (new) with **src/tests/end-to-end/utils/** (existing) ?

**Review by flomnes — CHANGES_REQUESTED** [review]
As requested by @sylvlecl, please create a single **testing_utils** CMake target.

File: src/tests/src/utils/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
+add_library(test_utils_unit
+        utils.cpp
+        utils.h
+        )
+
+target_include_directories(
+        test_utils_unit
+        PUBLIC
+        .
```
Comment (flomnes) [inline src/tests/src/utils/CMakeLists.txt]: ${CMAKE_CURRENT_SOURCE_DIR} ?

**Review by flomnes — CHANGES_REQUESTED** [review]

**Review by flomnes — APPROVED** [review]

**Pull request merged**